Add automatic docker build (and fix build)

modelrockettier opened the pull request:
There are 4 changes in this pull request:

1. Fix building `./...` from the go-dcpp root by removing the unused `filelist` module.  It was including the "go-dcpp/adc/types" module instead of "go-dc/adc/types" (it got moved, direct-connect/go-dc#21).

2. Add a Dockerfile and start script.

3. Add a GitHub Action to automatically build the go-dcpp docker container for every push and every pull request.

4. Fixed a nil pointer dereference in `ServeNMDC()`, exposed by the hub tests.

dennwc commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+WORKDIR /app
+
+# When building from the gohub repository, just copy it in
+copy . .
```
It may make sense to split the build into two parts: first get dependencies, then build the actual source. It should speed up consequent Docker image rebuilds:

```Dockerfile
# Copy go.mod and go.sum and get all the dependecies
COPY go.* ./ 
RUN go get -d -v ./...

# Build
COPY . .
# ... the rest of the file
```

dennwc commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+
+
+### Final container stage ###
+FROM alpine:latest
```
It's better to pin a specific version here. Sometimes Alpine updated Alpine images will fail to run Go binaries because those were build with a different musl version. It's probably a good idea to pin alpine version in the Go builder image too.

dennwc commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+EXPOSE 1411
+
+# The HubStats port (TCP), be careful exposing this to the internet
+EXPOSE 2112
```
Also 6060 for pprof with the same comment

dennwc commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+
+# Run this command periodically to make sure everything is working
+# Disabled by default since it clogs up the container logs
+#HEALTHCHECK CMD dcping ping 127.0.0.1:1411
```
OK, I think it makes sense to add a `-q` flag to the pinger so it won't spam the log.

modelrockettier replied:
The output from the pinger itself just gets discarded and never makes it to the container logs.

It's actually all the connects and disconnects done by the pinger that show up in the hub logs.

modelrockettier commented:
I've brought back the `filelist` package (and fixed it's import), pinned both the golang and final image's Alpine versions, split the go dependency retrieval into a separate docker layer, and exposed the pprof port.
